src/Microsoft.AspNet.Razor.Runtime/BufferedTextWriter.cs

```diff
@@ -0,0 +1,189 @@
+        public override void WriteLine()
+        {
+            BufferedStringCollection.Append(Environment.NewLine);
+        }
+
+        /// <inheritdoc />
+        public override void WriteLine(string value)
+        {
+            Write(value);
+            WriteLine();
+        }
+
+        /// <inheritdoc />
+        public override Task WriteLineAsync(char value)
+        {
+            WriteLine(value);
+            return _completedTask;
+        }
+
+        /// <inheritdoc />
+        public override Task WriteLineAsync(char[] value, int start, int offset)
+        {
+            WriteLine(value, start, offset);
+            return _completedTask;
+        }
+
+        /// <inheritdoc />
+        public override Task WriteLineAsync(string value)
+        {
+            WriteLine(value);
+            return _completedTask;
+        }
+
+        /// <inheritdoc />
+        public override Task WriteLineAsync()
+        {
+            WriteLine();
+            return _completedTask;
+        }
+        
+        /// <summary>
+        /// Copies the contents of the current <see cref="TextWriter"/> into another.
+        /// </summary>
+        /// <param name="writer">The <see cref="TextWriter"/> to which the content must be copied.</param>
+        public void CopyTo(TextWriter writer)
+        {
+            var targetStringCollectionWriter = writer as BufferedTextWriter;
+            if (targetStringCollectionWriter != null)
+            {
+                targetStringCollectionWriter.BufferedStringCollection.Buffer.Add(BufferedStringCollection.Buffer);
+            }
+            else
+            {
+                WriteList(writer, BufferedStringCollection.Buffer);
+            }
+        }
+
+        /// <summary>
+        /// Copies asynchronously the contents of the current <see cref="TextWriter"/> into another.
+        /// </summary>
+        /// <param name="writer">The <see cref="TextWriter"/> to which the content must be copied.</param>
+        public Task CopyToAsync(TextWriter writer)
+        {
+            var targetStringCollectionWriter = writer as BufferedTextWriter;
+            if (targetStringCollectionWriter != null)
+            {
+                targetStringCollectionWriter.BufferedStringCollection.Buffer.Add(BufferedStringCollection.Buffer);
+            }
+            else
+            {
+                return WriteListAsync(writer, BufferedStringCollection.Buffer);
+            }
+
+            return _completedTask;
+        }
+
+        /// <inheritdoc />
+        public override string ToString()
+        {
+            return BufferedStringCollection.ToString();
+        }
+
+        private static void WriteList(TextWriter writer, BufferEntryCollection values)
+        {
+            foreach (var value in values)
+            {
+                writer.Write(value);
+            }
+        }
+
+        private static async Task WriteListAsync(TextWriter writer, BufferEntryCollection values)
```

Comment:
Async's in this class are pretty much fake, no need to make separate call paths.